addons/sale_management/models/sale_order_template.py

```diff
@@ -29,6 +29,28 @@ def _get_default_require_payment(self):
         domain=[('model', '=', 'sale.order')],
         help="This e-mail template will be sent on confirmation. Leave empty to send nothing.")
     active = fields.Boolean(default=True, help="If unchecked, it will allow you to hide the quotation template without removing it.")
+    company_id = fields.Many2one('res.company', string='Company')
+
+    @api.constrains('company_id', 'sale_order_template_line_ids', 'sale_order_template_option_ids')
+    def _check_company_id(self):
+        for template in self:
+            companies = template.mapped('sale_order_template_line_ids.product_id.company_id') | template.mapped('sale_order_template_option_ids.product_id.company_id')
+            if len(companies) > 1:
+                raise ValidationError(_("Your template cannot contain products from multiple companies"))
+            elif companies and companies != template.company_id:
+                raise ValidationError((_("Your template contains products from company %s whereas your template belongs to company %s. \n Please change the company of your template or remove the products from other companies.") % (companies.mapped('display_name'), template.company_id.display_name)))
+
+    @api.onchange('sale_order_template_line_ids')
```

Comment:
The two onchange could be grouped to limit the company changes as much as possible:

```python
companies = self.mapped('sale_order_template_options_ids.product_id.company_id') | self.mapped('sale_order_template_line_ids.product_id.company_id')
if companies and not self.company_id in companies:
    self.company_id = companies[0]
```